### ISSUE TITLE
Missing "add-path-matcher" resource

### ISSUE THREAD

**Issue by hcliff** [hashicorp/terraform-provider-google#8662]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

The [compute_url_map](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_url_map) does a few different things, but they aren't exposed as individual entries, "path-matcher" and "host-rule" can only be created within a url map, despite there being [individual api endpoints for them](https://cloud.google.com/load-balancing/docs/url-map#gcloud_2) - see "add-path-matcher" and "add-host-rule".

Why does this matter? LB's are 1:1 with url maps, which means with this approach each workspace would need it's own LB, which for a  "coordination tool" is a bit odd, I'd expect multiple teams with discrete workspaces to be able to use the same LB. - If I'm misunderstanding this a point in the right direction would be appreciated!

### New or Affected Resource(s)

* https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_url_map

### Potential Terraform Configuration

```tf
resource "google_compute_url_map_path_matcher" "path_matcher" {
  url_map = "project/123/urlmaps/123"
  name = "allpaths"
  default_service = google_compute_backend_bucket.static.id
  path_rule {
      paths   = ["/home"]
      service = google_compute_backend_bucket.static.id
  }
}

resource "google_compute_url_map_host_rule" "host_rule"  {
  url_map = "project/123/urlmaps/123"
  hosts        = ["myothersite.com"]
  path_matcher = "otherpaths"
}
```

* #0000

**Comment by samclusker** [comment]
Super necessary - we're using the gcloud commands in a null_resource at the moment and it's very clunky, specifically around the create/destroy lifecycle of those resources

**Comment by pawelJas** [comment]
@hcliff There is no individual API to add a patch-matcher to UrlMap. `gcloud compute url-maps NAME add-path-matcher` is just a user friendly gcloud method to help build the correct payload, but it sends a regular [patch](https://cloud.google.com/compute/docs/reference/rest/v1/urlMaps/patch) request. This Patch method requires full `readwrite` and `use` IAM permission so not that easy to share between the teams. That gcloud command was added only because UrlMap is the most nested compute resource and therefore it was hard to design `gcloud compute url-maps create`  to allow all the possible configuration. General recommendation for UrlMaps is to use `gcloud compute url-maps import` which is very much aligned with current terraform model.

**Comment by hcliff** [comment]
Thanks, that's helpful context.

Smaller customers often reuse infrastructure, with looser multi-tenancy requirements, for example using one cloudsql instance with multiple databases within it. Here we were using "review apps" (a temporary deployment for every pull request) and didn't want to create one LB for each.

The above is annoying with the current design. The customer would need to re-implement `add-path-matcher` and `remove-path-matcher` to mutate a local file that they can then import.

**Comment by pawelJas** [comment]
Thank you @hcliff for explaining, but I am still not sure what exactly is needed.
Should there be a link to the "review apps" above?

I imagine something like below should happen:
- each workspace apply wants to have access to load-balancer/url-map
- apply should update the current url-map/load-balancer (add some new path-matcher)
- deleting the workspace (once temporary work is done) should remove the path-matcher.

I am not sure how splitting the `url-map` resource to `url-map` and `path-matcher` could help. the `url-map` should still reference the `path-matcher`.

The same can be resolved with the dynamic "path_matcher", no?

```
resource "google_compute_url_map" "default" {
  name            = "default"
  default_service = google_compute_backend_service.default["a-vm"].self_link

  dynamic "host_rule" {
    for_each = keys(local.instances)
    content {
      hosts        = ["${host_rule.value}.example.com"]
      path_matcher = host_rule.value
    }
  }

  dynamic "path_matcher" {
    for_each = keys(local.instances)
    content {
      name            = path_matcher.value
      default_service = google_compute_backend_service.default[path_matcher.value].self_link
    }
  }
```

do we have any example of similar

**Comment by hcliff** [comment]
I think we're on the right track, how does local.instances get populated?

**Comment by pawelJas** [comment]
Yes, local.instances would have to have all instances. You are right, it would require shared varriable. Getting it from the deployed URL_MAP sounds like what you mentioned about "re-implement add-path-matcher". 
Sure, I understand this is not ideal, but I cannot see how implementing new resource `google_compute_url_map_path_matcher` would be better. Particular workspace would still need to know the other created `google_compute_url_map_path_matcher` and this might be even harder, no?

**Comment by jgangemi** [comment]
any potential movement on this? i could also use this type of resource. we have a global load balancer that we need to register regional cloud run services with. my current work around is going to be clunky after hooks in terragrunt on apply for the following gcloud command:

```
gcloud compute url-maps add-path-matcher "${EXTERNAL_URL_MAP}" \
  --project="${HOST_PROJECT_ID}" \
  --global \
  --path-matcher-name="${SERVICE_NAME}" \
  --default-service="${SERVICE_NAME}-external-backend" \
  --new-hosts="${SERVICE_NAME}.external.${ENV}.example.com"
```

and a before hook w/ the remove command on destroy. `google_certificate_manager_certificate_map_entry` exists to allow this for certs being attached to the load balancer.
